### PR TITLE
Refactor internal shape and reduce reliance on Kurbo

### LINKED ISSUE #605: Refactor internal shape format to remove reliance on Kurbo
The data structure we are using to store paths, provided by the Kurbo library, doesn't provide a convenient way of manipulating shapes. This refactor should remove a direct reliance on Kurbo for the data structure. As part of this issue or later, we can work towards eliminating Kurbo as a dependency entirely since it will then be mostly unneeded (and fewer dependencies is better).

In the process of working toward improving the internal shape representation, the Overlay System needed to be separated from the shape definition. Overlays were tightly coupled to VectorShape / VectorPoint, we changed this relationship.

### LINKED ISSUE #616: Remove overlay ownership from VectorShape / VectorPoint
Currently overlays are tightly coupled to _VectorShape_ / _VectorPoint_, we want to change this relationship.

**New approach:** 
Send the _VectorShapes_ to an _OverlayRenderer_ and use an immediate mode renderer to render it.

This is part of moving toward this approach (WIP):
ShapeEditing (1)

### PR DESCRIPTION
**Pull request by otdavies**
This PR represents an **in progress** branch toward removing Kurbo as our primary shape representation. Overlays have been separated from the underlying vector representation to allow usage of VectorShape / VectorPoint as our internal shape representation.

# 🏁

- [x] Resolve overlay issues when multiple shapes are selected - (done I think @0HyperCube)
- [x] Resolve issue with losing overlays when childing layers to folders in certain situations.
- [x] Clean up overlays properly
- [x] Repair issues with text rendering.
- [x] Fix bug where part of path doesn't get rendered once you drag handles onto points.
- [x] Fix bug where the overlay handles don't get rendered if you drag a handle of a point (or use pen tool) (resolved @0HyperCube )
- [x] Don't snap against handle when dragging control point if handle locked onto control point
- [x] Fix regular polygon being centred without pressing <kbd>alt</kbd>
- [x] Fix multiple end paths causing paths to go to the wrong place (hover on text with multiple characters)
- [x] Fix Pen / repair pen tool (appears to work @0HyperCube)
- [x] Fix doc comment that's now failing for `fn intersects_quad(...)` at `graphene/src/layers/layer_info.rs:95`
- [x] Fix merge conflicts with `old-master`
- [ ] Allow users to get back their handle once they drag it onto a control point
- [ ] Treat all beziers as cubic and `None` handles as located at its anchor point, instead of cubic or linear
- [ ] Final clean up / comment pass.

And:
- ~~Remove kurbo from boolean operations~~ (another PR)
- ~~Implement bounding box behaviour with the new bézier library~~ (another PR)

Anyone is welcome to contribute, my availability to work on this is sparse currently.

Closes #605

### CODE REVIEW

**Comment by 0HyperCube** [comment]
> A layers uuid is not consistent throughout the layer's life time, counter to how UUIDs are supposed to work, aka it is replaced on move. This is not due to this refactor, this is how it is in live.

This is because we move layers by using the copy message them and the paste one - pasting should not keep the same layer ids because if you pasted the same layers twice this would cause a conflict.

**Comment by otdavies** [comment]
> > A layers uuid is not consistent throughout the layer's life time, counter to how UUIDs are supposed to work, aka it is replaced on move. This is not due to this refactor, this is how it is in live.
> 
> This is because we move layers by using the copy message them and the paste one - pasting should not keep the same layer ids because if you pasted the same layers twice this would cause a conflict.

This makes more sense in that light, I still think a layer (the first paste) should keep it's ID. It makes caching anything by layer uuid difficult. In the end this is really only needed for overlay rendering caching in my usage here. 

This quick overlay renderer solution is only a hold over until @TrueDoctor 's immediate mode renderer for overlays is more mature, so grain of salt about it's existing implementation. Thank you for helping out on this, it means a lot to me.

File: graphene/src/document.rs [hunk 1]
```diff
@@ -1,16 +1,16 @@
...
 use crate::layers::layer_info::{Layer, LayerData, LayerDataType};
 use crate::layers::shape_layer::ShapeLayer;
 use crate::layers::style::ViewMode;
 use crate::layers::text_layer::{Font, FontCache, TextLayer};
+use crate::layers::vector::vector_shape::VectorShape;
 use crate::{DocumentError, DocumentResponse, Operation};
 
 use glam::{DAffine2, DVec2};
-use kurbo::Affine;
+// use kurbo::Affine;
```
Comment (0HyperCube) [inline graphene/src/document.rs:13]: Can be removed.

File: graphene/src/document.rs [hunk 2]
```diff
@@ -100,7 +100,7 @@ impl Document {
 	}
 
 	/// Returns a mutable reference to the layer or folder at the path.
-	fn layer_mut(&mut self, path: &[LayerId]) -> Result<&mut Layer, DocumentError> {
+	pub fn layer_mut<'a>(&'a mut self, path: &'a [LayerId]) -> Result<&'a mut Layer, DocumentError> {
```
Comment (0HyperCube) [inline graphene/src/document.rs:103]: Why do we need the explicit lifetimes and the path to live as long as the resulting layer? (removing the explicit lifetimes compiles fine)

File: graphene/src/intersection.rs [hunk 3]
```diff
@@ -37,6 +38,24 @@ impl Quad {
...
+	/// Generates the axis aligned bounding box of the quad
+	pub fn bounding_box(&self) -> [DVec2; 2] {
+		[
+			self.0.into_iter().reduce(|a, b| a.min(b)).unwrap_or_default(),
+			self.0.into_iter().reduce(|a, b| a.max(b)).unwrap_or_default(),
+		]
+	}
+
+	/// Gets the centre of a quad
+	pub fn centre(&self) -> DVec2 {
```
Comment (0HyperCube) [inline graphene/src/intersection.rs:56]: British English.

File: graphene/src/layers/layer_info.rs [hunk 4]
```diff
@@ -86,7 +87,7 @@ pub trait LayerData {
 	/// let shape_id = 42;
 	/// let mut svg = String::new();
 	///
-	/// let quad = Quad::from_box([DVec2::ZERO, DVec2::ONE]);
+	/// let quad = Quad::from_box([DVec2::ZERO, DVec2::ONE*2.]);
```
Comment (0HyperCube) [inline graphene/src/layers/layer_info.rs:90]: This should probably actually be addressed.

File: editor/src/communication/graphite-test-document.graphite [hunk 5]
```diff
@@ -1 +1 @@
...
...5,1,181,14,49,49,14,254,78,11,47,47,11,1,178,14,49,0,1,0,36,255,69,3,199,2,43,0,34,0,99,64,22,32,31,28,27,24,23,20,19,16,15,12,11,10,13,3,2,9,1,1,3,2,74,75,176,50,80,88,64,25,0,0,3,0,82,6,4,2,2,2,25,75,7,5,2,3,3,1,94,0,1,1,24,1,76,27,64,25,6,4,2,2,3,2,131,0,0,3,0,82,7,5,2,3,3,1,94,0,1,1,24,1,76,89,64,11,19,19,19,19,19,21,20,18,8,7,28,43,5,6,7,35,52,39,38,39,33,53,55,17,39,53,33,21,7,17,51,17,39,53,33,21,7,17,51,17,39,53,33,21,7,17,51,3,192,12,7,62,3,11,3,252,198,85,83,1,13,83,218,84,1,9,79,219,84,1,10,80,102,9,127,51,10,24,147,6,47,11,1,178,14,49,49,14,254,75,1,181,14,49,49,14,254,75,1,181,14,49,49,14,254,77,0,0,1,0,51,255,77,2,132,2,43,0,27,0,91,64,20,24,23,22,19,18,15,14,11,10,9,10,3,2,25,8,2,1,3,2,74,75,176,50,80,88,64,23,0,0,1,0,132,4,1,2,2,25,75,0,3,3,1,94,5,1,1,1,24,1,76,27,64,23,4,1,2,3,2,131,0,0,1,0,132,0,3,3,1,94,5,1,1,1,24,1,76,89,64,9,21,19,19,21,34,18,6,7,26,43,4,6,21,35,52,38,39,35,53,55,17,39,53,51,21,7,17,51,17,39,53,51,21,7,17,23,21,35,1,147,14,64,21,2,250,70,71,248,75,246,74,244,67,71,240,2,163,14,13,163,3,49,10,1,177,14,49,49,14,254,75,1,181,14,49,49,14,254,79,10,49,0,2,0,34,255,246,2,38,2,43,0,20,0,29,0,145,64,15,19,18,15,14,4,0,3,29,13,12,3,4,5,2,74,75,176,29,80
```
[the remark this answers is not in the document]
Comment (Keavon) [reply editor/src/communication/graphite-test-document.graphite:1]: Yeah, we should just make sure this test file never includes a font when we manually generate it. (Or ideally, we'd move towards being able to manually run a test which regenerates the file, something along the lines of the idea proposed in [this blog post](https://matklad.github.io/2022/03/26/self-modifying-code.html)).

[the remark this answers is not in the document]
Comment (0HyperCube) [reply editor/src/communication/graphite-test-document.graphite:1]: @Keavon No files now have fonts embeded in them as of #659

File: .vscode/settings.json [hunk 6]
```diff
@@ -46,4 +46,5 @@
 	"html.format.wrapLineLength": 200,
 	"files.eol": "\n",
 	"files.insertFinalNewline": true,
+	"rust-analyzer.procMacro.attributes.enable": true,
```
[the remark this answers is not in the document]
Comment (otdavies) [reply .vscode/settings.json:49]: Yeah, I believe so. Remove if preferred.

File: editor/src/document/document_message_handler.rs [hunk 7]
```diff
@@ -160,26 +160,26 @@ impl DocumentMessageHandler {
...
+	// 		// TODO: Create VectorManipulatorShape when creating a kurbo shape as a stopgap, rather than on each new selection
+	// 		match &layer.ok()?.data {
+	// 			LayerDataType::Shape(shape) => Some(VectorShape::new(path_to_shape.to_vec(), viewport_transform, &shape.path, shape.closed, responses)),
+	// 			LayerDataType::Text(text) => Some(VectorShape::new(path_to_shape.to_vec(), viewport_transform, &text.to_bez_path_nonmut(font_cache), true, responses)),
+	// 			_ => None,
+	// 		}
+	// 	});
+
+	// 	shapes.collect::<Vec<VectorShape>>()
+	// }
```
[the remark this answers is not in the document]
Comment (otdavies) [reply editor/src/document/document_message_handler.rs]: Yes.

File: editor/src/input/input_mapper.rs [hunk 8]
```diff
@@ -103,9 +103,11 @@ impl Default for Mapping {
 			// Path
 			entry! {action=PathToolMessage::DragStart { add_to_selection: KeyShift }, key_down=Lmb},
 			entry! {action=PathToolMessage::PointerMove { alt_mirror_angle: KeyAlt, shift_mirror_distance: KeyShift }, message=InputMapperMessage::PointerMove},
+			entry! {action=PathToolMessage::SelectPoint, message=InputMapperMessage::DoubleClick},
```
[the remark this answers is not in the document]
Comment (otdavies) [reply editor/src/input/input_mapper.rs]: Not used I believe.

File: editor/src/viewport_tools/snapping.rs [hunk 9]
```diff
@@ -52,10 +53,11 @@ impl SnapOverlays {
...
 					if is_axis {
-						Operation::AddOverlayLine {
+						Operation::AddLine {
 							path: layer_path.clone(),
 							transform,
 							style: style::PathStyle::new(Some(Stroke::new(COLOR_ACCENT, 1.0)), style::Fill::None),
+							insert_index: -1,
 						}
 					} else {
 						Operation::AddOverlayEllipse {
```
[the remark this answers is not in the document]
Comment (otdavies) [reply editor/src/viewport_tools/snapping.rs]: Likely just missed this, feel free to convert.

File: graphene/src/document.rs (hunk 2 above)
Comment (otdavies) [reply graphene/src/document.rs]: Yep, likely residual, removing is fine.

**Pull request merged**